Self hosted parser completion

Hejsil opened the pull request:
This is in no way done, but I thought, that instead of just pushing my work on the self-hosted parser to master, I would make this PR. This allows me to write a lot of tests that fail, and slowly work towards getting them all passing.

Checklist!
- [x] `extern "c" fn puts(s: &const u8) c_int;`
- [x] Muti line strings
- [x] Indexing
- [x] Slicing
- [x] Types
  - [x] Array
  - [x] Slice
  - [x] Function
  - [x] `error`
- [x] Literals
  - [x] Boolean
  - [x] Char
  - [x] `null`
- [x] `unreachable`
- [x] `this`
- [x] Operator precedence and expression grouping (parentheses)
- [x] Struct declaration
- [x] Enum declaration
- [x] Union declaration
- [x] Error set declaration
- [x] Container initializers
- [x] Labeled blocks
- [x] `switch`
- [x] `while`
- [x] `for`
- [x] `if`
- [x] `defer`
- [x] `catch`
- [x] `comptime`
- [x] Inline assembly
- [x] Coroutines
- [x] `use`
- [ ] Parse `std`

When all these are checked and the self-hosted parser can parse the entire std library and all tests, then it's ready for a merge.

andrewrk commented:
Great work so far. It's been really exciting to watch your progress

andrewrk commented on std/zig/parser.zig:
```diff
@@ -1730,4 +1737,145 @@ test "zig fmt" {
+
+    try testCanonical(
+        \\test "test array" {
+        \\    const a : [2]u8 = [2]u8{ 1, 2 };
```
I think canonical var decl is:
`const name: type = value;`
(No space between colon and var name)

bnoordhuis commented on std/zig/parser.zig:
```diff
@@ -1730,4 +1737,145 @@ test "zig fmt" {
+    );
+
+    try testCanonical(
+        \\test "percendence" {
```
'precedence' typo?

Hejsil commented:
I've now added test cases for what I hope is all of Zig's syntax. I've also added a checklist with all everything that is missing from the self-hosted parser.

andrewrk commented on std/zig/parser.zig:
```diff
@@ -1731,3 +1750,504 @@ test "zig fmt" {
+        \\    x += 1;
+        \\    suspend |p| {
+        \\    }
+        \\    const p = async simpleAsyncFn() cache unreachable;
```
I hate it when the cache is unreachable :)

Hejsil replied:
Lol, yes :)

Hejsil commented:
@andrewrk The variable declaration syntax in the self-hosted parser is currently not being parsed correctly. `var a: u8 = 1;` is parsed into`var a: ( u8 = 1 ) ;`.

In the C++ compiler, this is solved by parsing type expressions as prefix operators first and therefore requiring parentheses for infix operators to be used in type expressions. This solution cannot be used in the self-hosted parser because of the way expressions are parsed.

Do we rewrite expression parsing to make this work or disallow assignment as an expression?

Hejsil commented:
There is also the option of checking if the type expression is an assignment and then transform `var a: (u8 = 0);` -> `var a: u8 = 0;`. The only problem with this solution is, that we are using an arena as an allocator, so we cannot free the NodeInfixOp node.

Hejsil commented:
Also, in the C++ compiler, this is allowed:
```
var a: u8 = 0;
const b = a = 2;
```

I don't think this has any use. `@typeOf(b) == void`, and I don't think that is what a person intended if they wrote this piece of code.

Hejsil commented:
This is also a problem when implementing initializers :/
```
fn main() void {}
```

Gets parsed as:
```
fn main() (void {})
```

andrewrk commented:
I think assignments should not be expressions.

Hejsil commented:
Alright. Then we just need to solve #760 :)

Hejsil commented on std/zig/parser.zig:
```diff
@@ -125,6 +125,48 @@ pub const Parser = struct {
+        Optional: Parser,
+
+        /// Optional can be reverted by adding the Required state to the stack.
+        Required,
```
I don't know if I like this Optional and Required idea. It a footgun in so many ways, because we have no way of reverting the side effects other states have done. This means that `return 5 + ;` would actually store `5` as `return`s rhs. In this case that isn't a problem because there is an actual syntax error, but I'm sure there are corner cases I haven't thought of.

andrewrk commented:
From the appveyor build:
`lld: error: undefined symbol: memmove`

Wow, I've not seen this happen yet. I guess we need to add a `memmove` implementation to std/special/builtin.zig.

andrewrk commented:
All tests passing, merged!

Let's open new issues for the new things.

tiehuis commented:
Great work, @Hejsil.
